File: odom_equalizer/odom_equalizer.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy
import gz.transport13   as ign_transport
from gz.msgs10.pose_pb2 import Pose
from gz.msgs10.boolean_pb2 import Boolean
from nav_msgs.msg import Odometry
import sys


class OdomEqualizer(Node):
    """
    This node subscribes to /sim/Odometry (ROS2) and forwards the pose
    to Ignition Transport via the /world/sim/set_pose service.
    """

    def __init__(self):
        super().__init__('odom_equalizer')

        # QoS: reliable, (optionally) transient-local if latching is needed
        qos_profile = QoSProfile(
            depth=10,
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.VOLATILE  # or TRANSIENT_LOCAL if desired
        )

        # Subscribe to ROS2 Odometry topic
        self.subscriber = self.create_subscription(
            Odometry,
            '/sim/Odometry',
            self._odom_callback,
            qos_profile
        )

        # Ignition Transport client (named differently from the ROS2 node)
        self.ign_node = ign_transport.Node()

        # Keep the latest Odometry message; we'll push it to Ignition at 10 Hz
        self._latest_odom = None

        # Timer to call Ignition service at up to 10 Hz
        timer_period = 0.1  # seconds (10 Hz)
        self.timer = self.create_timer(timer_period, self._timer_send_pose)

        self.get_logger().info('OdomEqualizer node has been started')

    def _odom_callback(self, msg: Odometry):
        """
        Receiver callback for ROS2 Odometry. Store the latest message for the timer.
        """
        # Log that it received a new Odometry message
        self.get_logger().info(f'Received Odometry: {msg.pose.pose.position.x}, '
                                f'{msg.pose.pose.position.y}, {msg.pose.pose.position.z}')
        self._latest_odom = msg
        

    def _timer_send_pose(self):
        """
        Timer callback (10 Hz). If we've received at least one Odometry,
        convert it to an Ignition Pose and send the service request.
        """
        # log the timer event        
        if self._latest_odom is None:
            self.get_logger().info('No Odometry data received yet, skipping Ignition pose update')
            return
        
        self.get_logger().info('Timer triggered to send pose to Ignition')

        odom = self._latest_odom

        # Build the Ignition Pose message
        pose_msg = Pose()
        pose_msg.name = 'vehicle_blue'
        pose_msg.position.x = odom.pose.pose.position.x
        pose_msg.position.y = odom.pose.pose.position.y
        pose_msg.position.z = odom.pose.pose.position.z
        pose_msg.orientation.x = odom.pose.pose.orientation.x
        pose_msg.orientation.y = odom.pose.pose.orientation.y
        pose_msg.orientation.z = odom.pose.pose.orientation.z
        pose_msg.orientation.w = odom.pose.pose.orientation.w

        # Send the service request (blocking up to 2000 ms)

        try:
            outcome, body  = self.ign_node.request(
                service='/world/sim/set_pose',  # make sure the world name matches your simulator
                request=pose_msg,
                request_type=Pose,
                response_type=Boolean,
                timeout=2000  # in milliseconds
            )
        except Exception as e:
            self.get_logger().error(f"Ignition request threw exception: {e}")
            return

        # If the request couldn't even be delivered
        if outcome is False:
            self.get_logger().error('Failed to dispatch set_pose request to Ignition')
            # self.get_logger().error(f'The body of the response was: {body}')
            return
    
        # If we reach here, the pose was successfully set
        elif outcome is True:
            self.get_logger().info(
                f'The requeste returned outcome: {outcome}'
                f'Successfully set pose at '
                f'(x={pose_msg.position.x:.2f}, y={pose_msg.position.y:.2f}, z={pose_msg.position.z:.2f})'
            )

    def destroy_node(self):
        """
        Override destroy_node to cleanly close the Ignition client as well.
        """
        super().destroy_node()
        try:
            self.ign_node.close()
            self.get_logger().info('Closed Ignition Transport node')
        except AttributeError:
            # If `close` is not available, ignore
            pass


def main(args=None):
    rclpy.init(args=sys.argv)
    node = OdomEqualizer()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info('KeyboardInterrupt received, shutting down')
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: odom_equalizer/src/odom_equalizer.cpp
#include "odom_equalizer.hpp"
#include "gz/math/Vector3.hh"

odom_equalizer::OdomEqualizer::OdomEqualizer() : Node("odom_equalizer")
{
    this->inizialize();
}

nav_msgs::msg::Odometry get_initial_odom()
{
    auto odom = nav_msgs::msg::Odometry();
    ignition::math::Quaterniond correction_quat(
        ignition::math::Vector3d(M_PI_2, 0, 0)  // Roll +90 degrees
    );
    odom.pose.pose.orientation.w = correction_quat.W();
    odom.pose.pose.orientation.x = correction_quat.X();
    odom.pose.pose.orientation.y = correction_quat.Y();
    odom.pose.pose.orientation.z = correction_quat.Z();
    odom.pose.pose.position.x = 0.0;
    odom.pose.pose.position.y = 0.0;
    odom.pose.pose.position.z = -0.30;
    return odom;
}


void odom_equalizer::OdomEqualizer::inizialize()
{
 
    this->load_parameters();

    this->odomSub = this->create_subscription<nav_msgs::msg::Odometry>(this->param_odom_topic, 1, std::bind(&OdomEqualizer::odom_callback, this, std::placeholders::_1));

    this->ign_node = std::make_shared<ignition::transport::Node>();

    // Initialize lates_odom with a default value to avoid the publishing of an uninitialized message
    this->latest_odom = get_initial_odom();

    // create a timer to periodically call set_pose
    this->timer = this->create_wall_timer(std::chrono::milliseconds(100), std::bind(&OdomEqualizer::set_pose, this));

    RCLCPP_INFO(this->get_logger(), "NODE IS INITIALIZED");
}

void odom_equalizer::OdomEqualizer::load_parameters()
{
    this->declare_parameter<std::string>("odom_topic", "/Odometry");
    this->param_odom_topic = this->get_parameter("odom_topic").get_value<std::string>();

    this->declare_parameter<std::string>("service_name", "/world/sim/set_pose");
    this->param_service_topic = this->get_parameter("service_name").get_value<std::string>();

    RCLCPP_INFO(this->get_logger(), "PARAMETERS LOADED");
}

void odom_equalizer::OdomEqualizer::odom_callback(const nav_msgs::msg::Odometry::SharedPtr msg)
{
    RCLCPP_INFO(this->get_logger(), "RECEIVED ODOMETRY MESSAGE");
    this->latest_odom = *msg;
}

void odom_equalizer::OdomEqualizer::set_pose()
{
    ignition::msgs::Pose pose_msg;
    pose_msg.set_name("vehicle_blue");

    // Set position
    auto position = pose_msg.mutable_position();
    position->set_x(this->latest_odom.pose.pose.position.x);
    position->set_y(this->latest_odom.pose.pose.position.y);
    position->set_z(this->latest_odom.pose.pose.position.z + 0.8); // this offset is needed, otherwise the vehicle will be partially underground

    // Set orientation
    ignition::math::Quaterniond quat(
        this->latest_odom.pose.pose.orientation.w, 
        this->latest_odom.pose.pose.orientation.x, 
        this->latest_odom.pose.pose.orientation.y, 
        this->latest_odom.pose.pose.orientation.z
    );

    ignition::math::Quaterniond correction_quat(
        ignition::math::Vector3d(-M_PI_2, 0, 0)  // Roll -90 degrees
    );

    quat = quat * correction_quat;

    auto orientation = pose_msg.mutable_orientation();

    orientation->set_x(quat.X());
    orientation->set_y(quat.Y());
    orientation->set_z(quat.Z());
    orientation->set_w(quat.W());

    ignition::msgs::Boolean reply;
    bool call_result = false;
    // Call Ignition service
    bool ok = ign_node->Request<ignition::msgs::Pose, ignition::msgs::Boolean>("/world/sim/set_pose", pose_msg, 100, reply, call_result);

    if (ok && call_result && reply.data()) 
    {
        RCLCPP_INFO(this->get_logger(), "SUCCESS");
    }
    else 
    {
        RCLCPP_INFO(this->get_logger(), "FAILED");
    }
}